src/views/CanvasOrderbook/index.tsx

```diff
@@ -0,0 +1,297 @@
+          : [];
+      asksSlice = [...emptyAskRows, ...asks.reverse()];
+
+      let bidsSlice: Array<PerpetualMarketOrderbookLevel | undefined> = [];
+      const emptyBidRows =
+        bids.length < maxRowsPerSide
+          ? new Array<undefined>(maxRowsPerSide - bids.length).fill(undefined)
+          : [];
+      bidsSlice = [...bids, ...emptyBidRows];
+
+      return {
+        asksSlice,
+        bidsSlice,
+      };
+    }, [asks, bids]);
+
+    const orderbookRef = useRef<HTMLDivElement>(null);
+    useCenterOrderbook({ orderbookRef, marketId: currentMarket });
+
+    /**
+     * Display top or bottom spreadRow when center spreadRow is off screen
+     */
+    const spreadRowRef = useRef<HTMLDivElement>(null);
+
+    const displaySide = useSpreadRowScrollListener({
+      orderbookRef,
+      spreadRowRef,
+    });
+
+    /**
+     * Row action
+     */
+    const currentInput = useSelector(getCurrentInput);
+    const dispatch = useDispatch();
+    const onRowAction = useCallback(
+      (price: Nullable<number>) => {
+        console.log(price);
+        if (currentInput === 'trade' && price) {
+          dispatch(setTradeFormInputs({ limitPriceInput: price?.toString() }));
+        }
+      },
+      [currentInput]
+    );
+
+    const { canvasRef: asksCanvasRef } = useDrawOrderbook({
+      data: [...asksSlice].reverse(),
+      histogramRange,
+      histogramSide,
+      side: 'ask',
+    });
+
+    const { canvasRef: bidsCanvasRef } = useDrawOrderbook({
+      data: bidsSlice,
+      histogramRange,
+      histogramSide,
+      side: 'bid',
+    });
+
+    return (
+      <Styled.OrderbookContainer ref={ref}>
+        <Styled.OrderbookContent $isLoading={!hasOrderbook}>
+          <Styled.Header>
+            <span>
+              {stringGetter({ key: STRING_KEYS.SIZE })} {id && <Tag>{id}</Tag>}
+            </span>
+            <span>
+              {stringGetter({ key: STRING_KEYS.PRICE })} <Tag>USD</Tag>
+            </span>
+            <span>{stringGetter({ key: STRING_KEYS.MINE })}</span>
+          </Styled.Header>
+
+          {displaySide === 'top' && (
+            <Styled.SpreadRow
+              side="top"
+              spread={spread}
+              spreadPercent={spreadPercent}
+              tickSizeDecimals={tickSizeDecimals}
+            />
+          )}
+
+          <Styled.OrderbookWrapper ref={orderbookRef}>
+            {!hasOrderbook ? (
+              <LoadingSpace id="canvas-orderbook" />
+            ) : (
+              <>
+                <Styled.OrderbookSideContainer $side="asks">
+                  <Styled.HoverRows $bottom>
+                    {asksSlice.map((row: PerpetualMarketOrderbookLevel | undefined, idx) =>
+                      row ? (
+                        <Styled.Row
+                          key={idx}
+                          title={`${row.price}`}
+                          onClick={
+                            row?.price
```

Comment:
out of curiosity, when does a row not have a price (assuming that the row is a `PerpetualMarketOrderbookLevel` and not `undefined`)?

Reply:
price is `Nullable<number>` coming form Abacus

Reply:
can we assume `row` is non null here since this is in `row ? (X)`?